src/zkstar.rs

```diff
@@ -0,0 +1,160 @@
+struct ZStarNBuilder {
+    modulus: BigNumber,
+}
+#[allow(unused)]
+impl ZStarNBuilder {
+    pub fn new(n: BigNumber) -> Self {
+        ZStarNBuilder { modulus: n }
+    }
+    /// One of two ways of constructing an element of [ZStarN] by validating that an unverifier instance of [ZStarNUnverified]
+    /// is properly constructed with respect to the current [ZStarNBuilder]
+    /// (the other way is by randomly sampling the element in the multiplicative group modulo n)
+    pub fn validate(&self, unverified: ZStarNUnverified) -> Result<ZStarN, InternalError> {
+        if unverified.value().is_zero() {
+            return Err(InternalError::IsZero);
+        } else if unverified.value() > self.modulus() {
+            return Err(InternalError::LargerThanModulus);
+        } else if unverified.value() < &BigNumber::zero() {
+            return Err(InternalError::NegativeElement);
+        }
+        let result = unverified.value().gcd(self.modulus());
+        if result != BigNumber::one() {
+            return Err(InternalError::NotCoprime);
+        }
+        Ok(ZStarN {
+            value: unverified.value().to_owned(),
+            builder: self,
+        })
+    }
+    /// Return the modulus of [ZStarNBuilder]
+    fn modulus(&self) -> &BigNumber {
+        &self.modulus
+    }
+}
+
+/// A multiplicative group of elements modulo `N` as defined by [ZStarNBuilder]
+/// Elements of this group are in the interval (0, N) and validated to be as such.
+#[derive(Serialize)]
+struct ZStarN<'a> {
+    value: BigNumber,
+    builder: &'a ZStarNBuilder,
+}
+#[allow(unused)]
+impl<'a> ZStarN<'a> {
+    /// Return the value stored in [ZStarN]
+    fn as_bignumber(&self) -> &BigNumber {
+        &self.value
+    }
+    fn serialize(&self) -> Result<Vec<u8>, InternalError> {
+        serialize!(self.as_bignumber())
+    }
+    /// Randomly samples an element of the multiplicative group modulo N (as defined by the builder)
+    /// This is one of two ways of constructing an element [ZStarN]
+    /// (the other way is by validating an instance of [ZStarNUnverified])
+    fn random_element<R: RngCore + CryptoRng>(
+        builder: &'a ZStarNBuilder,
+        rng: &mut R,
+    ) -> Result<Self, InternalError> {
+        utils::random_bn_in_z_star(rng, builder.modulus()).map(|value| Self { value, builder })
+    }
+}
+/// Unverified, deserialized value that claims to be in the multiplicative
+/// group of integers mod `N`, for some modulus `N`.
+///
+/// This can be verified into a valid [`ZStarN`] using a [`ZStarNBuilder`].
+#[derive(Deserialize)]
+struct ZStarNUnverified {
+    value: BigNumber,
+}
+#[allow(unused)]
+impl ZStarNUnverified {
+    fn new(value: BigNumber) -> Self {
+        ZStarNUnverified { value }
+    }
+    fn value(&self) -> &BigNumber {
+        &self.value
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::paillier::PaillierDecryptionKey;
+    use crate::zkstar::*;
+
+    #[test]
+    fn test_zkstar_verification() {
```

Comment:
suggestion: Are you open to calling this something more descriptive? This is a personal pet peeve more than anything, but I don't like having `test` in the name of tests for rust because they're already flagged with `#[test]`. 

I prefer names like `zkstar_constructor_rejects_elements_outside_group` or `zkstar_verification_works`.

Reply:
Love the new names, thank you!